### PR TITLE
Upload refactoring + shadow drive support

### PR DESCRIPTION
**Pull request by febo**
Improved the way that new upload methods can be implemented:

- [x] Creation of a new `Uploader` trait to be implemented by all upload methods.
- [x] Threading logic was moved from each upload method to the `ParallelUpload` trait. This avoids code duplication and simplifies the implementation.
- [x] Moved upload method implementation under a new methods sub-module. New methods only (or ideally) need to add/modify files from this module.
- [x] Submit `edit` requests for existing files in shadow drive storage
- [x] Refactor nft.storage upload to improve performance (uploading batches of files)
- [x] Merge 'main' branch into PR branch

### CODE REVIEW

**Review by danenbm — APPROVED** [review]

File: src/upload/uploader.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        Ok(errors)
+    }
+}
+
+/// Factory function for uploader objects. This function returns an initializes uploader trait object.
+///
+/// The storage uploader is selected based on the configuration `UploadMethod`.
+pub async fn initialize(
```
Comment (danenbm) [inline src/upload/uploader.rs:162]: This is cool!

File: src/upload/methods/aws.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+};
+use crate::{common::*, config::*};
+
+pub struct AWSMethod {
+    pub aws_client: Arc<Client>,
+    pub bucket: String,
+}
+
+impl AWSMethod {
+    pub async fn initialize(config_data: &ConfigData) -> Result<AWSMethod> {
```
[the remark this answers is not in the document]
Comment (febo) [reply src/upload/methods/aws.rs:19]: Good point - I made the suggested changes.

File: src/upload/methods/bundlr.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            .get(format!("{node}/price/solana/{data_size}"))
+            .send()
+            .await?
+            .text()
+            .await?
+            .parse::<u64>()?;
+        Ok(required_amount)
+    }
+
+    async fn send(
```
[the remark this answers is not in the document]
Comment (febo) [reply src/upload/methods/bundlr.rs:183]: Interesting alternative - thanks!

File: src/upload/methods/aws.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl ParallelUploader for AWSMethod {
+    fn upload_asset(&self, asset_info: AssetInfo) -> JoinHandle<Result<(String, String)>> {
+        let client = self.aws_client.clone();
+        let bucket = self.bucket.clone();
+        tokio::spawn(async move { AWSMethod::send(client, bucket, asset_info).await })
```
[the remark this answers is not in the document]
Comment (danenbm) [reply src/upload/methods/aws.rs:79]: Ah yes I forgot that the Result would make it back via the `JoinHandle`, wasn't looking closely enough.

**Comment by samuelvanderwaal** [comment]
Missing the `shdw` option [here](https://github.com/metaplex-foundation/sugar/blob/04b334f2bb23605d95749e69d9c6fcd6fe7d9157/src/create_config/process.rs#L565).

**Comment by febo** [comment]
Closed in favour of #242

**Pull request closed without merging**